// flow-parser/dist/transform/print/print.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.print = print;
var _mutateESTreeASTForPrettier = _interopRequireDefault(require("../../utils/mutateESTreeASTForPrettier"));
var prettier = _interopRequireWildcard(require("prettier"));
var _comments = require("./comments/comments");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
async function print(ast, originalCode, prettierOptions = {}, visitorKeys) {
  const program = ast;
  if (program.body.length === 0) {
    var _program$docblock;
    const docblockComment = (_program$docblock = program.docblock) == null ? void 0 : _program$docblock.comment;
    if (docblockComment != null) {
      return '/*' + docblockComment.value + '*/\n';
    }
    return '';
  }
  const codeForPrinting = (0, _comments.mutateESTreeASTCommentsForPrettier)(program, originalCode);
  (0, _mutateESTreeASTForPrettier.default)(program, visitorKeys);
  let pluginParserName = 'flow';
  let pluginParser;
  let pluginPrinter;
  try {
    const prettierHermesPlugin = await Promise.resolve().then(() => _interopRequireWildcard(require('prettier-plugin-hermes-parser')));
    pluginParser = prettierHermesPlugin.parsers.hermes;
    pluginPrinter = prettierHermesPlugin.printers;
    pluginParserName = 'hermes';
  } catch {
    const prettierFlowPlugin = require('prettier/plugins/flow');
    pluginParser = prettierFlowPlugin.parsers.flow;
  }
  return prettier.format(codeForPrinting, {
    ...prettierOptions,
    parser: pluginParserName,
    requirePragma: false,
    plugins: [{
      parsers: {
        [pluginParserName]: {
          ...pluginParser,
          parse() {
            return program;
          }
        }
      },
      printers: pluginPrinter
    }]
  });
}
